polkadot/xcm/pallet-xcm/src/lib.rs

```diff
@@ -1194,13 +1212,41 @@ impl<T: Config> QueryHandler for Pallet<T> {
 }
 
 impl<T: Config> Pallet<T> {
+	/// Validate `assets` to be reserve-transferred and return their reserve location.
+	fn validate_assets_and_find_reserve(
+		assets: &[MultiAsset],
+		dest: &MultiLocation,
+	) -> Result<TransferType, Error<T>> {
+		let mut reserve = None;
+		for asset in assets.iter() {
+			if let Fungible(x) = asset.fun {
+				// If fungible asset, ensure non-zero amount.
+				ensure!(!x.is_zero(), Error::<T>::Empty);
+			}
+			let transfer_type =
+				<T::XcmExecutor as AssetTransferSupport>::determine_for(&asset, dest)
+					.map_err(Error::<T>::InvalidAsset)?;
+			// Ensure asset is not teleportable to `dest`.
+			ensure!(transfer_type != TransferType::Teleport, Error::<T>::Filtered);
+			if let Some(reserve) = reserve.as_ref() {
+				// Ensure transfer for multiple assets uses same reserve location (only fee may have
+				// different reserve location)
+				ensure!(reserve == &transfer_type, Error::<T>::TooManyReserves);
+			} else {
+				// asset reserve identified
+				reserve = Some(transfer_type);
+			}
+		}
+		reserve.ok_or(Error::<T>::Empty)
+	}
+
 	fn do_reserve_transfer_assets(
```

Comment:
Do you think we also need to change weight formula for the `reserve_asset_transfer` (and `limited_reserve_transfer_assets`) calls? Right now it assumes that we are executing single local `SetFeesMode + TransferReserveAsset` XCM program + adds weight, computed by benchmarking. And after this PR, we may execute up to `3` such programs

Reply:
Actually - I'm not sure why the current formula is `T::WeightInfo::reserve_transfer_assets().saturating_add(<weight-of-xcm-that-is-executed-locally>)`, so I'm not sure about my proposal as well :) Like there's a benchmark for `reserve_asset_transfer` and the `reserve_asset_transfer` executes this message, so it looks like it already shall include the weight of the local XCM execution. Maybe some dummy XCM executor is supposed to be used in `pallet-xcm` benchmarks...In any case - most probably we need to triple the effect of `<weight-of-xcm-that-is-executed-locally>`, because that's the worst case now

Reply:
👍 yes, we need to rework the benchmark to test on the heaviest scenario